File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/null_object/solution/get_number_of_uses_of_variable_expression_visitor.cpp
// get_number_of_uses_of_variable_expression_visitor.cpp

#include <string>
#include "expression_visitor.h"
#include "get_number_of_uses_of_variable_expression_visitor.h"
#include "variable.h"

get_number_of_uses_of_variable_expression_visitor::get_number_of_uses_of_variable_expression_visitor(const std::string & a_name)
: my_name(a_name), my_number_of_uses_of_variable(0)
{
}

void get_number_of_uses_of_variable_expression_visitor::visit(variable & a_variable)
{
	if (a_variable.get_name() == my_name)
	{
		++my_number_of_uses_of_variable;
	}
}

int get_number_of_uses_of_variable_expression_visitor::get_number_of_uses_of_variable() const
{
	return my_number_of_uses_of_variable;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/null_object/lab/GetNumberOfUsesOfVariableExpressionVisitor.java
// GetNumberOfUsesOfVariableExpressionVisitor.java

public class GetNumberOfUsesOfVariableExpressionVisitor extends ExpressionVisitor
{
	public GetNumberOfUsesOfVariableExpressionVisitor(String aName)
	{
		myName = aName;
		myNumberOfUsesOfVariable = 0;
	}

	public void visit(Variable aVariable)
	{
		if (aVariable.getName().equals(myName))
		{
			++myNumberOfUsesOfVariable;
		}
	}

	public int getNumberOfUsesOfVariable()
	{
		return myNumberOfUsesOfVariable;
	}

	private String myName;
	private int myNumberOfUsesOfVariable;
}
